File: algorithms/code/leetcode/lc328_odd_even_linked_list/LC328OddEvenLinkedList.java
package leetcode.lc328_odd_even_linked_list;

/**
 * https://leetcode.com/problems/odd-even-linked-list/
 * #medium
 */
public final class LC328OddEvenLinkedList {
    public static class ListNode {
        int val;
        ListNode next;

        ListNode(final int x) {
            val = x;
        }
    }

    public ListNode oddEvenList(final ListNode headValue) {
        ListNode head = headValue;
        ListNode evenHead = new ListNode(0);
        ListNode oddHead = new ListNode(0);
        ListNode evenTail = evenHead;
        ListNode oddTail = oddHead;
        boolean odd = true;
        while (head != null) {
            if (odd) {
                oddTail.next = head;
                oddTail = oddTail.next;
                odd = false;
            } else {
                evenTail.next = head;
                evenTail = evenTail.next;
                odd = true;
            }
            head = head.next;
        }
        evenTail.next = null;
        oddTail.next = evenHead.next;
        return oddHead.next;
    }
}
package leetcode.lc328_odd_even_linked_list;

import org.junit.jupiter.api.Test;

import leetcode.lc328_odd_even_linked_list.LC328OddEvenLinkedList.ListNode;

import static org.junit.jupiter.api.Assertions.assertEquals;

public final class LC328OddEvenLinkedListTests {
    @Test
    public void testExample() throws Exception {
        ListNode head = new ListNode(0);
        ListNode tail = head;
        for (int i = 1; i <= 5; i++) {
            tail.next = new ListNode(i);
            tail = tail.next;
        }
        ListNode result = new LC328OddEvenLinkedList().oddEvenList(head.next);
        int[] expected = { 1, 3, 5, 2, 4 };
        for (int value : expected) {
            assertEquals(value, result.val);
            result = result.next;
        }
        assertEquals(null, result);
    }

    @Test
    public void testNull() throws Exception {
        assertEquals(null, new LC328OddEvenLinkedList().oddEvenList(null));
    }
}

File: algorithms/code/leetcode/lc328_odd_even_linked_list/lc328_odd_even_linked_list.py
# https://leetcode.com/problems/odd-even-linked-list/

from __future__ import annotations

import unittest


class ListNode:
    def __init__(
        self, val: int = 0, following: ListNode | None = None
    ) -> None:
        self.val = val
        self.next = following


class Solution:
    def oddEvenList(self, head: ListNode | None) -> ListNode | None:
        evenHead = ListNode(0)
        oddHead = ListNode(0)
        evenTail = evenHead
        oddTail = oddHead
        odd = True
        while head is not None:
            if odd:
                oddTail.next = head
                oddTail = oddTail.next
                odd = False
            else:
                evenTail.next = head
                evenTail = evenTail.next
                odd = True
            head = head.next
        evenTail.next = None
        oddTail.next = evenHead.next
        return oddHead.next


class TestCode(unittest.TestCase):
    def test_example(self) -> None:
        head = ListNode(0)
        tail = head
        for i in range(1, 6):
            tail.next = ListNode(i)
            tail = tail.next
        result = Solution().oddEvenList(head.next)
        expected = [1, 3, 5, 2, 4]
        for value in expected:
            assert result is not None
            assert value == result.val
            result = result.next
        assert result is None

    def test_null(self) -> None:
        assert Solution().oddEvenList(None) is None
